lib/rpush/daemon/fcm/delivery.rb

```diff
@@ -0,0 +1,139 @@
+module Rpush
+  module Daemon
+    module Fcm
+      # https://firebase.google.com/docs/cloud-messaging/server
+      class Delivery < Rpush::Daemon::Delivery
+        include MultiJsonHelper
+
+        PROJECT_ID = Settings.firebase.fcm_push_notification.project_id
+        ENDPOINT = "https://fcm.googleapis.com/v1/projects/#{PROJECT_ID}/messages:send".freeze
+
+        def initialize(app, http, notification, batch)
+          @app = app
+          @http = http
+          @notification = notification
+          @batch = batch
+
+          @uri = URI.parse(ENDPOINT)
+        end
+
+        def perform
+          handle_response(do_post)
+        rescue SocketError => error
+          mark_retryable(@notification, Time.now + 10.seconds, error)
+          raise
+        rescue StandardError => error
+          mark_failed(error)
+          raise
+        ensure
+          @batch.notification_processed
+        end
+
+        protected
+
+        def handle_response(response)
+          case response.code.to_i
+          when 200
+            ok
+          when 400
+            bad_request(response)
+          when 401
+            unauthorized
```

Comment:
As we roll this out, it's worth catching up with the 400/401s we've been facing from Firebase in our web push. As far as I remember, there was a bit of intermittent nature to them and I don't know if Voice ever got to the bottom of it

https://housecall.atlassian.net/browse/VOICE-6463

At the very least we'll need to monitor closely to know if we're failing to deliver more than normal on FCM v1 compared to legacy.

Reply:
> it's worth catching up with the 400/401s

What do you mean by "catch up with"? As in have visibility into why they happen? We'll ensure we have some monitoring before we roll out.

@nat-aryucharoen Since failures get logged to sumologic we can add some reporting there. If needed, we can also add some metrics to datadog with the Reflections API, but I figured the sumologic logs should be good enough.

Reply:
might be this? for 401: https://firebase.google.com/docs/reference/fcm/rest/v1/ErrorCode
> THIRD_PARTY_AUTH_ERROR (HTTP error code = 401) APNs certificate or web push auth key was invalid or missing. | A message targeted to an iOS device or a web push registration could not be sent. Check the validity of your development and production credentials